File: ios/ClawChatKit/Sources/ClawChatKit/Pairing/PairingManager.swift
import Foundation

/// Result of a successful pairing.
public struct PairingResult: Sendable {
    public let deviceToken: String
    public let gatewayId: String
}

/// Result of a successful reconnection.
public struct ReconnectResult: Sendable {
    public let gatewayId: String
    public let gatewayOnline: Bool
    public let newDeviceToken: String?
}

/// Manages pairing and reconnection flows over a WebSocket connection.
public actor PairingManager {
    private let client: WebSocketClient
    private let pairingTimeout: TimeInterval = 30

    public init(client: WebSocketClient) {
        self.client = client
    }

    /// Pair with a relay using a 6-character pairing code.
    /// Sends `app.pair` and awaits `app.paired` or `app.pair.error`.
    public func pair(code: String, deviceName: String) async throws -> PairingResult {
        let message = AppPair(pairingCode: code, deviceName: deviceName)
        await client.send(message)

        let timeout = pairingTimeout
        let messages = await client.messages

        return try await withThrowingTaskGroup(of: PairingResult.self) { group in
            group.addTask {
                for await msg in messages {
                    switch msg {
                    case .appPaired(let paired):
                        return PairingResult(deviceToken: paired.deviceToken, gatewayId: paired.gatewayId)
                    case .appPairError(let error):
                        throw Self.pairingError(from: error.error)
                    case .error(let error):
                        if error.code == .unauthorized {
                            throw PairingError.unauthorized
                        }
                    default:
                        continue
                    }
                }
                throw PairingError.networkError("Connection closed before pairing completed")
            }

            group.addTask {
                try await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                throw PairingError.timeout
            }

            let result = try await group.next()!
            group.cancelAll()
            return result
        }
    }

    /// Reconnect to the relay using a stored device token.
    /// Sends `app.connect` and awaits `app.connected` or error.
    public func reconnect(deviceToken: String) async throws -> ReconnectResult {
        let message = AppConnect(deviceToken: deviceToken)
        await client.send(message)

        let timeout = pairingTimeout
        let messages = await client.messages

        return try await withThrowingTaskGroup(of: ReconnectResult.self) { group in
            group.addTask {
                for await msg in messages {
                    switch msg {
                    case .appConnected(let connected):
                        return ReconnectResult(
                            gatewayId: connected.gatewayId,
                            gatewayOnline: connected.gatewayOnline ?? false,
                            newDeviceToken: connected.newDeviceToken
                        )
                    case .error(let error):
                        if error.code == .unauthorized {
                            throw PairingError.unauthorized
                        }
                        throw PairingError.networkError(error.message)
                    default:
                        continue
                    }
                }
                throw PairingError.networkError("Connection closed before reconnection completed")
            }

            group.addTask {
                try await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                throw PairingError.timeout
            }

            let result = try await group.next()!
            group.cancelAll()
            return result
        }
    }

    // MARK: - Helpers

    private static func pairingError(from reason: PairErrorReason) -> PairingError {
        switch reason {
        case .invalidCode:
            return .invalidCode
        case .codeExpired:
            return .codeExpired
        case .expired:
            return .codeExpired
        case .gatewayOffline:
            return .gatewayOffline
        }
    }
}

File: clawchat/android/clawchat-kit/src/main/kotlin/com/clawchat/kit/pairing/PairingManager.kt
package com.clawchat.kit.pairing

import com.clawchat.kit.protocol.*
import com.clawchat.kit.websocket.WebSocketClient
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withTimeout

/** Result of a successful pairing. */
data class PairingResult(
    val deviceToken: String,
    val gatewayId: String,
)

/** Result of a successful reconnection. */
data class ReconnectResult(
    val gatewayId: String,
    val gatewayOnline: Boolean,
    val newDeviceToken: String? = null,
)

/**
 * Manages pairing and reconnection flows over a WebSocket connection.
 */
class PairingManager(
    private val client: WebSocketClient,
    private val timeoutMs: Long = 30_000,
) {
    /** Pair with a relay using a 6-character pairing code. */
    suspend fun pair(code: String, deviceName: String): PairingResult {
        val message = AppPair(pairingCode = code, deviceName = deviceName)
        client.send(message)

        return withTimeout(timeoutMs) {
            client.messages.first { msg ->
                when (msg) {
                    is ClawChatMessage.AppPairedMsg -> true
                    is ClawChatMessage.AppPairErrorMsg -> true
                    is ClawChatMessage.ErrorMsg ->
                        msg.value.code is ClawChatErrorCode.Unauthorized
                    else -> false
                }
            }.let { msg ->
                when (msg) {
                    is ClawChatMessage.AppPairedMsg ->
                        PairingResult(msg.value.deviceToken, msg.value.gatewayId)
                    is ClawChatMessage.AppPairErrorMsg ->
                        throw pairingException(msg.value.error)
                    is ClawChatMessage.ErrorMsg ->
                        throw PairingException.Unauthorized()
                    else -> error("unreachable")
                }
            }
        }
    }

    /** Reconnect to the relay using a stored device token. */
    suspend fun reconnect(deviceToken: String): ReconnectResult {
        val message = AppConnect(deviceToken = deviceToken)
        client.send(message)

        return withTimeout(timeoutMs) {
            client.messages.first { msg ->
                when (msg) {
                    is ClawChatMessage.AppConnectedMsg -> true
                    is ClawChatMessage.ErrorMsg -> true
                    else -> false
                }
            }.let { msg ->
                when (msg) {
                    is ClawChatMessage.AppConnectedMsg ->
                        ReconnectResult(
                            gatewayId = msg.value.gatewayId,
                            gatewayOnline = msg.value.gatewayOnline ?: false,
                            newDeviceToken = msg.value.newDeviceToken,
                        )
                    is ClawChatMessage.ErrorMsg -> {
                        if (msg.value.code is ClawChatErrorCode.Unauthorized) {
                            throw PairingException.Unauthorized()
                        }
                        throw PairingException.NetworkError(msg.value.message)
                    }
                    else -> error("unreachable")
                }
            }
        }
    }

    private fun pairingException(reason: PairErrorReason): PairingException = when (reason) {
        PairErrorReason.INVALID_CODE -> PairingException.InvalidCode()
        PairErrorReason.CODE_EXPIRED -> PairingException.CodeExpired()
        PairErrorReason.EXPIRED -> PairingException.CodeExpired()
        PairErrorReason.GATEWAY_OFFLINE -> PairingException.GatewayOffline()
    }
}
